pkcs11/_pkcs11.pyx

```diff
@@ -1187,16 +1194,50 @@ cdef class lib:
         # to keep a pointer to the C_GetFunctionList address returned by dlsym()
         cdef C_GetFunctionList_ptr C_GetFunctionList
 
-        handle = dlfcn.dlopen(so.encode('utf-8'), dlfcn.RTLD_LAZY | dlfcn.RTLD_LOCAL)
-        if handle == NULL:
-            raise RuntimeError(dlfcn.dlerror())
+        IF UNAME_SYSNAME == "Windows":
+            self._handle = mswin.LoadLibraryW(so)
+            if self._handle == NULL:
+                raise RuntimeError("Cannot open library at {}: {}".format(path, self._winerrormsg()))
 
-        C_GetFunctionList = <C_GetFunctionList_ptr> dlfcn.dlsym(handle, 'C_GetFunctionList')
-        if C_GetFunctionList == NULL:
-            raise RuntimeError("{} is not a PKCS#11 library: {}".format(so, dlfcn.dlerror()))
+            if self._handle != NULL:
+                C_GetFunctionList = <C_GetFunctionList_ptr> mswin.GetProcAddress(self._handle, 'C_GetFunctionList')
+                if C_GetFunctionList == NULL:
+                    raise RuntimeError("{} is not a PKCS#11 library: {}".format(so, self._winerrormsg()))
+        ELSE:
+            self._handle = dlfcn.dlopen(so.encode('utf-8'), dlfcn.RTLD_LAZY | dlfcn.RTLD_LOCAL)
+            if self._handle == NULL:
+                raise RuntimeError(dlfcn.dlerror())
+
+            C_GetFunctionList = <C_GetFunctionList_ptr> dlfcn.dlsym(self._handle, 'C_GetFunctionList')
+            if C_GetFunctionList == NULL:
+                raise RuntimeError("{} is not a PKCS#11 library: {}".format(so, dlfcn.dlerror()))
 
         assertRV(C_GetFunctionList(&_funclist))
 
+
+    cdef _unload_pkcs11_lib(self):
+        """Unload a PKCS#11 library.
+
+        This method will dynamically unload a PKCS11 library.
+
+        This is a private method, and must never be called directly.
+        Called when a lib instance is destroyed.
+        """
+
+        IF UNAME_SYSNAME == "Windows":
+            if self._handle != NULL:
+                mswin.FreeLibrary(self._handle)
+        ELSE:
+            if self._handle != NULL:
+                dlfcn.dlclose(self._handle)
+
+
+    IF UNAME_SYSNAME == "Windows":
```

Comment:
Should this go in your `mswin.pxd`?

Reply:
I am assuming you mean the `cdef _winerrormsg()` code section surrounded with the `IF` statement.
I have a personal preference to put only headers declaration into `.pxd` files; Cython however allows code sections into these files. Fair enough, I'll move this as per request.